### DISCUSSION TITLE
Getting JuliaFormatter to work with helix editor

### DISCUSSION THREAD

**Discussion by haxfn** [helix-editor/helix#11510] [Troubleshooting]
When doing
```bash
$ hx --health julia  # I get:
Configured language servers:
  ✓ julia: /home/vishal/.juliaup/bin/julia
Configured debug adapter: None
Configured formatter: None
Highlight queries: ✓
Textobject queries: ✓
Indent queries: ✓
```

So, I would like to setup a formatter for julia. In this case I am using [JuliaFormatter](https://github.com/domluna/JuliaFormatter.jl).

When doing
```bash
$ julia -e 'using JuliaFormatter; format(ARGS[1]);' src/main.jl
```
The file is successfully formatted automatically.

But editing helix-config/languages.toml, to include:
```toml
[[language]]
name = "julia"
formatter = { command = "julia", timeout = 60, args = [ "-e", "using JuliaFormatter; format(ARGS[1]);", ] }
```
And manually running helix::format inside helix on the same file, I get helix-log julia error:
```
2024-08-17T11:32:06.020 helix_view::document [ERROR] Formatter error: ERROR: BoundsError: attempt to access 0-element Vector{String} at index [1]
Stacktrace:
 [1] getindex(A::Vector{String}, i1::Int64)
   @ Base ./essentials.jl:13
 [2] top-level scope
   @ none:1
```

Which hints that the formatter does not recieve the path of the file as an argument? How should I go about adding the said JuliaFormatter to helix? What is expected off of the formatter command? What is passed to the formatter command?

**Comment by haxfn** [comment]
possibly related: https://github.com/helix-editor/helix/issues/3596

**Reply by haxfn** [reply]
Maybe STDIN needs to be read.

**Reply by the-mikedavis** [reply]
External formatters are passed the file contents as stdin and are expected to print the formatter doc as stdout

**Comment by sschuldenzucker** [comment]
FWIW I found the following to work fine:

```
formatter = { command = "julia", timeout = 60, args = [ "--startup-file=no", "--history-file=no", "--project", "--quiet", "-e", "using JuliaFormatter; print(format_text(read(stdin, String)));", ] }
```

(passing `--project` is a matter of taste but I do it so I can control the version of JuliaFormatter to use for my local project)

**Comment by mauro3** [comment]
x-ref: https://github.com/domluna/JuliaFormatter.jl/issues/868
